### PR TITLE
fix: Eventarc post deploy process

### PR DESCRIPTION
**Pull request by davenportjw**
- Moves the eventarc creation process to separate resources
- Eventarc now triggers workflows
- Adds workflow for additional future customization of post terraform deployment
- Adds wait steps where needed for IAM role syncing

### CODE REVIEW

File: modules/data_warehouse/main.tf [hunk 1]
```diff
@@ -55,17 +56,62 @@ resource "random_id" "id" {
...
 resource "google_service_account" "cloud_function_service_account" {
   project      = module.project-services.project_id
   account_id   = "cloud-function-sa-${random_id.id.hex}"
   display_name = "Service Account for Cloud Function Execution"
 }
 
-# TODO: scope this down
-resource "google_project_iam_member" "cloud_function_service_account_editor_role" {
+# # Grant the Functions service account Object Create / Delete access
+resource "google_project_iam_member" "cloud_function_service_account_storage_role" {
```
Comment (bharathkkb) [inline modules/data_warehouse/main.tf:68]: Can we use `google_storage_bucket_iam_member` to scope it to just a bucket?

File: modules/data_warehouse/main.tf [hunk 2]
```diff
@@ -55,17 +56,62 @@ resource "random_id" "id" {
...
   project = module.project-services.project_id
-  role    = "roles/editor"
+  role    = "roles/storage.objectAdmin"
+  member  = "serviceAccount:${google_service_account.cloud_function_service_account.email}"
+
+  depends_on = [
+    google_service_account.cloud_function_service_account
+  ]
```
Comment (bharathkkb) [inline modules/data_warehouse/main.tf:75]: nit: explicit dependency is not necessary as email is a computed attrib which make this as implicit dependency

File: modules/data_warehouse/main.tf [hunk 3]
```diff
@@ -504,25 +502,79 @@ resource "google_cloudfunctions2_function" "function" {
...
+  role    = "roles/run.invoker"
+  member  = "serviceAccount:${google_service_account.workflow_service_account.email}"
+
+  depends_on = [
+    google_service_account.workflow_service_account
+  ]
+}
+
+# # Grant the Workflow service account ability to run as other account
+resource "google_project_iam_member" "workflow_service_account_act_as_role" {
```
Comment (bharathkkb) [inline modules/data_warehouse/main.tf:541]: We should scope this to just the SA it needs access to using `google_service_account_iam_member`

File: modules/data_warehouse/main.tf [hunk 4]
```diff
@@ -531,7 +583,82 @@ resource "google_storage_bucket_object" "startfile" {
...
+# # Grant the Eventar service account Workflow Invoker Access
+resource "google_project_iam_member" "eventarc_service_account_invoke_role" {
+  project = module.project-services.project_id
+  role    = "roles/workflows.invoker"
+  member  = "serviceAccount:${google_service_account.eventarc_service_account.email}"
+
+  depends_on = [
+    google_service_account.eventarc_service_account
+  ]
+}
```
Comment (bharathkkb) [inline modules/data_warehouse/main.tf:664]: nit: newline

**Review by bharathkkb — APPROVED** [review]
Some minor comments

**Comment by davenportjw** [comment]
Feedback captured to incorporate for next PR

**Pull request merged**